File: swift/Sources/MoneyHero/YahooFinanceClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public struct YahooHTTPResponse: Sendable {
    public let requestURL: URL
    public let statusCode: Int
    public let data: Data

    public init(requestURL: URL, statusCode: Int, data: Data) {
        self.requestURL = requestURL
        self.statusCode = statusCode
        self.data = data
    }
}

public enum YahooFinanceClientError: Error, Sendable {
    case invalidURL(String)
    case nonHTTPResponse
    case missingSeedCookies
    case invalidCrumb
    case stockChartUnauthorized(statusCode: Int)
    case unexpectedStatusCode(statusCode: Int, url: URL)
}

public actor YahooFinanceClient {
    public static let sp500Symbol = "^GSPC"
    public static let eurUsdSymbol = "EURUSD=X"
    public static let gbpUsdSymbol = "GBPUSD=X"
    public static let goldSymbol = "GC=F"

    private let userAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile Safari/604.1"
    private let sessionTTL: TimeInterval = 30 * 60

    private let session: URLSession
    private let manualRedirectSession: URLSession
    private var yahooSession: YahooSession?

    public init(session: URLSession = .shared) {
        self.session = session

        let configuration = URLSessionConfiguration.ephemeral
        configuration.httpCookieStorage = nil
        configuration.httpShouldSetCookies = false
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        self.manualRedirectSession = URLSession(
            configuration: configuration,
            delegate: RedirectBlockerDelegate(),
            delegateQueue: nil
        )
    }

    public func fetchSP500Chart() async throws -> YahooHTTPResponse {
        let symbol = try normalizeSymbol(Self.sp500Symbol)
        return try await fetchStockChartNormalized(symbol: symbol)
    }

    public func fetchStockChart(symbol rawSymbol: String) async throws -> YahooHTTPResponse {
        let symbol = try normalizeSymbol(rawSymbol)
        return try await fetchStockChartNormalized(symbol: symbol)
    }

    public func fetchEURUSDChart() async throws -> YahooHTTPResponse {
        try await fetchCurrencyChart(symbol: Self.eurUsdSymbol)
    }

    public func fetchGBPUSDChart() async throws -> YahooHTTPResponse {
        try await fetchCurrencyChart(symbol: Self.gbpUsdSymbol)
    }

    public func fetchGoldQuoteChart() async throws -> YahooHTTPResponse {
        var components = URLComponents(string: "https://query1.finance.yahoo.com/v8/finance/chart/\(Self.goldSymbol)")
        components?.queryItems = [
            URLQueryItem(name: "range", value: "1d"),
            URLQueryItem(name: "interval", value: "1d")
        ]

        let url = try makeURL(from: components, raw: "gold quote chart")
        return try await performRequest(url: url, includeUserAgent: true)
    }

    public func fetchGoldHistoryChart() async throws -> YahooHTTPResponse {
        let range = stockPeriodRange()

        var components = URLComponents(string: "https://query1.finance.yahoo.com/v8/finance/chart/\(Self.goldSymbol)")
        components?.queryItems = [
            URLQueryItem(name: "period1", value: String(range.period1)),
            URLQueryItem(name: "period2", value: String(range.period2)),
            URLQueryItem(name: "interval", value: "1d")
        ]

        let url = try makeURL(from: components, raw: "gold history chart")
        return try await performRequest(url: url, includeUserAgent: true)
    }

    private func fetchStockChartNormalized(symbol: String) async throws -> YahooHTTPResponse {
        let firstSession = try await ensureSession()
        let firstResponse = try await performStockChartRequest(symbol: symbol, yahooSession: firstSession)

        if firstResponse.statusCode == 401 || firstResponse.statusCode == 403 {
            yahooSession = nil
            let reminted = try await ensureSession(forceRemint: true)
            let retryResponse = try await performStockChartRequest(symbol: symbol, yahooSession: reminted)

            if retryResponse.statusCode == 401 || retryResponse.statusCode == 403 {
                throw YahooFinanceClientError.stockChartUnauthorized(statusCode: retryResponse.statusCode)
            }

            try validateStatusCode(retryResponse)
            return retryResponse
        }

        try validateStatusCode(firstResponse)
        return firstResponse
    }

    private func fetchCurrencyChart(symbol: String) async throws -> YahooHTTPResponse {
        var components = URLComponents(string: "https://query1.finance.yahoo.com/v8/finance/chart/\(symbol)")
        components?.queryItems = [
            URLQueryItem(name: "range", value: "2d"),
            URLQueryItem(name: "interval", value: "1d")
        ]

        let url = try makeURL(from: components, raw: "currency chart")
        let response = try await performRequest(url: url, includeUserAgent: true)
        try validateStatusCode(response)
        return response
    }

    private func performStockChartRequest(symbol: String, yahooSession: YahooSession) async throws -> YahooHTTPResponse {
        let range = stockPeriodRange()

        var components = URLComponents(string: "https://query1.finance.yahoo.com/v8/finance/chart/\(symbol)")
        components?.queryItems = [
            URLQueryItem(name: "period1", value: String(range.period1)),
            URLQueryItem(name: "period2", value: String(range.period2)),
            URLQueryItem(name: "interval", value: "1d"),
            URLQueryItem(name: "events", value: "div"),
            URLQueryItem(name: "crumb", value: yahooSession.crumb)
        ]

        let url = try makeURL(from: components, raw: "stock chart")
        return try await performRequest(url: url, cookieHeader: yahooSession.cookieHeader, includeUserAgent: true)
    }

    private func ensureSession(forceRemint: Bool = false) async throws -> YahooSession {
        if !forceRemint, let existing = yahooSession, Date().timeIntervalSince(existing.createdAt) < sessionTTL {
            return existing
        }

        let minted = try await mintSession()
        yahooSession = minted
        return minted
    }

    private func mintSession() async throws -> YahooSession {
        let cookieHeader = try await fetchSeedCookieHeader()
        let crumb = try await fetchCrumb(cookieHeader: cookieHeader)
        return YahooSession(cookieHeader: cookieHeader, crumb: crumb, createdAt: Date())
    }

    private func fetchSeedCookieHeader() async throws -> String {
        guard let url = URL(string: "https://fc.yahoo.com/") else {
            throw YahooFinanceClientError.invalidURL("https://fc.yahoo.com/")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        let (_, response) = try await manualRedirectSession.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw YahooFinanceClientError.nonHTTPResponse
        }

        var headerFields: [String: String] = [:]
        for (key, value) in httpResponse.allHeaderFields {
            guard let keyString = key as? String else {
                continue
            }
            guard let valueString = value as? String else {
                continue
            }
            headerFields[keyString] = valueString
        }

        let cookies = HTTPCookie.cookies(withResponseHeaderFields: headerFields, for: url)
        let header = cookies
            .map { "\($0.name)=\($0.value)" }
            .joined(separator: "; ")

        guard !header.isEmpty else {
            throw YahooFinanceClientError.missingSeedCookies
        }

        return header
    }

    private func fetchCrumb(cookieHeader: String) async throws -> String {
        guard let url = URL(string: "https://query2.finance.yahoo.com/v1/test/getcrumb") else {
            throw YahooFinanceClientError.invalidURL("https://query2.finance.yahoo.com/v1/test/getcrumb")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue(cookieHeader, forHTTPHeaderField: "Cookie")
        request.setValue(userAgent, forHTTPHeaderField: "User-Agent")

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw YahooFinanceClientError.nonHTTPResponse
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            throw YahooFinanceClientError.unexpectedStatusCode(statusCode: httpResponse.statusCode, url: url)
        }

        let crumb = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !crumb.isEmpty, crumb != "null" else {
            throw YahooFinanceClientError.invalidCrumb
        }

        return crumb
    }

    private func performRequest(url: URL, cookieHeader: String? = nil, includeUserAgent: Bool) async throws -> YahooHTTPResponse {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"

        if let cookieHeader {
            request.setValue(cookieHeader, forHTTPHeaderField: "Cookie")
        }
        if includeUserAgent {
            request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
        }

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw YahooFinanceClientError.nonHTTPResponse
        }

        return YahooHTTPResponse(requestURL: url, statusCode: httpResponse.statusCode, data: data)
    }

    private func makeURL(from components: URLComponents?, raw: String) throws -> URL {
        guard let url = components?.url else {
            throw YahooFinanceClientError.invalidURL(raw)
        }

        return url
    }

    private func validateStatusCode(_ response: YahooHTTPResponse) throws {
        guard (200..<300).contains(response.statusCode) else {
            throw YahooFinanceClientError.unexpectedStatusCode(statusCode: response.statusCode, url: response.requestURL)
        }
    }

    private func stockPeriodRange(referenceDate: Date = Date()) -> (period1: Int, period2: Int) {
        let period2 = Int(referenceDate.timeIntervalSince1970)
        let twoYearsAndPaddingDays: TimeInterval = 744 * 24 * 60 * 60
        let period1 = Int(referenceDate.addingTimeInterval(-twoYearsAndPaddingDays).timeIntervalSince1970)
        return (period1: period1, period2: period2)
    }
}

private struct YahooSession: Sendable {
    let cookieHeader: String
    let crumb: String
    let createdAt: Date
}

private final class RedirectBlockerDelegate: NSObject, URLSessionTaskDelegate {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

File: flutter/lib/market_data/yahoo_finance_client.dart
import "dart:async";
import "dart:convert";

import "package:http/http.dart" as http;

import "../domain/market_normalization.dart";
import "market_data_protocols.dart";

class YahooFinanceClientError implements Exception {
  const YahooFinanceClientError(this.code);

  final String code;
}

class YahooFinanceClient implements MarketDataProviding {
  YahooFinanceClient({http.Client? httpClient}) : _httpClient = httpClient ?? http.Client();

  static const String sp500Symbol = "^GSPC";
  static const String goldSymbol = "GC=F";

  static const String _userAgent =
      "Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Mobile Safari/604.1";
  static const Duration _sessionTtl = Duration(minutes: 30);

  final http.Client _httpClient;
  _YahooSession? _session;

  @override
  Future<YahooHttpResponse> fetchStockChart({required String symbol}) async {
    final String normalized = normalizeSymbol(symbol);
    _YahooSession session = await _ensureSession();
    YahooHttpResponse first = await _fetchStockChartWithSession(symbol: normalized, session: session);
    if (first.statusCode == 401 || first.statusCode == 403) {
      _session = null;
      session = await _ensureSession(forceRemint: true);
      final YahooHttpResponse retry = await _fetchStockChartWithSession(symbol: normalized, session: session);
      if (retry.statusCode == 401 || retry.statusCode == 403) {
        throw const YahooFinanceClientError("stockChartUnauthorized");
      }
      _validateStatusCode(retry);
      return retry;
    }
    _validateStatusCode(first);
    return first;
  }

  @override
  Future<YahooHttpResponse> fetchCurrencyChart({required String displayedCurrency}) async {
    final String currency = normalizeCurrencyCode(displayedCurrency);
    final Uri url = Uri.https(
      "query1.finance.yahoo.com",
      "/v8/finance/chart/${currency}USD=X",
      <String, String>{"range": "2d", "interval": "1d"},
    );
    final YahooHttpResponse response = await _performRequest(url: url, includeUserAgent: true);
    _validateStatusCode(response);
    return response;
  }

  @override
  Future<YahooHttpResponse> fetchGoldQuoteChart() async {
    final Uri url = Uri.https(
      "query1.finance.yahoo.com",
      "/v8/finance/chart/$goldSymbol",
      <String, String>{"range": "1d", "interval": "1d"},
    );
    final YahooHttpResponse response = await _performRequest(url: url, includeUserAgent: true);
    _validateStatusCode(response);
    return response;
  }

  @override
  Future<YahooHttpResponse> fetchGoldHistoryChart() async {
    final ({int period1, int period2}) range = _stockPeriodRange();
    final Uri url = Uri.https(
      "query1.finance.yahoo.com",
      "/v8/finance/chart/$goldSymbol",
      <String, String>{"period1": "${range.period1}", "period2": "${range.period2}", "interval": "1d"},
    );
    final YahooHttpResponse response = await _performRequest(url: url, includeUserAgent: true);
    _validateStatusCode(response);
    return response;
  }

  Future<YahooHttpResponse> _fetchStockChartWithSession({required String symbol, required _YahooSession session}) {
    final ({int period1, int period2}) range = _stockPeriodRange();
    final Uri url = Uri.https("query1.finance.yahoo.com", "/v8/finance/chart/$symbol", <String, String>{
      "period1": "${range.period1}",
      "period2": "${range.period2}",
      "interval": "1d",
      "events": "div",
      "crumb": session.crumb,
    });
    return _performRequest(url: url, includeUserAgent: true, cookieHeader: session.cookieHeader);
  }

  Future<_YahooSession> _ensureSession({bool forceRemint = false}) async {
    if (!forceRemint && _session != null && _session!.expiresAt.isAfter(DateTime.now())) {
      return _session!;
    }
    final String cookie = await _mintCookieHeader();
    final String crumb = await _mintCrumb(cookieHeader: cookie);
    final _YahooSession session = _YahooSession(
      cookieHeader: cookie,
      crumb: crumb,
      expiresAt: DateTime.now().add(_sessionTtl),
    );
    _session = session;
    return session;
  }

  Future<String> _mintCookieHeader() async {
    final http.Request request = http.Request("GET", Uri.parse("https://fc.yahoo.com/"));
    request.followRedirects = false;
    request.maxRedirects = 0;
    final http.StreamedResponse response = await _httpClient.send(request);
    final List<String> setCookies = response.headers.entries
        .where((MapEntry<String, String> entry) => entry.key.toLowerCase() == "set-cookie")
        .map((MapEntry<String, String> entry) => entry.value)
        .toList();
    final List<String> values = setCookies
        .map((String cookie) => cookie.split(";").first.trim())
        .where((String cookie) => cookie.isNotEmpty)
        .toList();
    if (values.isEmpty) {
      throw const YahooFinanceClientError("missingSeedCookies");
    }
    return values.join("; ");
  }

  Future<String> _mintCrumb({required String cookieHeader}) async {
    final Uri url = Uri.parse("https://query2.finance.yahoo.com/v1/test/getcrumb");
    final YahooHttpResponse response = await _performRequest(
      url: url,
      includeUserAgent: true,
      cookieHeader: cookieHeader,
    );
    _validateStatusCode(response);
    final String crumb = utf8.decode(response.data).trim();
    if (crumb.isEmpty || crumb == "null") {
      throw const YahooFinanceClientError("invalidCrumb");
    }
    return crumb;
  }

  Future<YahooHttpResponse> _performRequest({required Uri url, required bool includeUserAgent, String? cookieHeader}) async {
    final Map<String, String> headers = <String, String>{};
    if (includeUserAgent) {
      headers["User-Agent"] = _userAgent;
    }
    if (cookieHeader != null) {
      headers["Cookie"] = cookieHeader;
    }
    final http.Response response = await _httpClient.get(url, headers: headers);
    return YahooHttpResponse(requestUrl: url, statusCode: response.statusCode, data: response.bodyBytes);
  }

  void _validateStatusCode(YahooHttpResponse response) {
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw YahooFinanceClientError("unexpectedStatus:${response.statusCode}");
    }
  }

  ({int period1, int period2}) _stockPeriodRange() {
    final int period2 = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    const int twoYearsAndPaddingDays = 744 * 24 * 60 * 60;
    return (period1: period2 - twoYearsAndPaddingDays, period2: period2);
  }
}

class _YahooSession {
  const _YahooSession({required this.cookieHeader, required this.crumb, required this.expiresAt});

  final String cookieHeader;
  final String crumb;
  final DateTime expiresAt;
}
